### PR TITLE
Improve zsh integration: rc options, loading order, clash avoidance

### PR DESCRIPTION
**Pull request by romkatv**
This commit does the following to improve zsh integration:

- Ensures that `rcs` and `global_rcs` shell options are respected.
- Ensures that all global and user rc files are sources in the right order.
- Avoids clashes with user-defined functions.
- Avoids clashes with user-defined aliases, including global aliases.
- Speeds up zsh startup by removing an invocation of `env`.
- Fixes a bug in parsing of environment variables with` \n` in them.

See https://youtrack.jetbrains.com/issue/IDEA-245380.

One might think that the bug mentioned above could be fixed by adding all zsh rc files to `plugins/terminal/resources` (`.zshenv`, `.zshrc`, `.zlogin`, etc.) and sourcing the original user rc files from them. This, however, wouldn't work if any of the user rc files unset option `rcs` or change `ZDOTDIR` parameter.

This code ensures correct IDEA code injection as long as `/etc/zshenv` doesn't unset `rcs` or change `ZDOTDIR`. `/etc/zshenv` rarely does such things. If it does, injecting code into zsh startup sequence by exporting `ZDOTDIR` prior to invoking `zsh` becomes impossible.

### CODE REVIEW

**Comment by segrey** [comment]
@romkatv Thank you for the PR! Have you submitted CLA? If no, please visit https://www.jetbrains.com/agreements/cla/ and follow the steps to sign the CLA electronically.

**Comment by romkatv** [comment]
@sergey Signed.

**Comment by segrey** [comment]
@romkatv Thanks. Hope I will have time to take a look at it next week.

**Comment by pete-woods** [comment]
Would be really nice to have this taken care of. The existing zsh init script doesn’t work at all if your project path has a space in anywhere.

File: plugins/terminal/resources/.zshenv [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+# Stop right here if this is not interactive shell.
+[[ -o 'interactive' ]] || 'builtin' 'return' '0'
+
+'builtin' 'typeset' '-i' '_jedi_restore_aliases'
+if [[ -o 'aliases' ]]; then
+  'builtin' 'setopt' 'no_aliases'
+  # No need to quote everything after this point because aliases
+  # are disabled. We didn't do this earlier because we must source
+  # the user's .zshenv with the default shell options.
+  _jedi_restore_aliases=1
```
[the remark this answers is not in the document]
Comment (romkatv) [reply plugins/terminal/resources/.zshenv:34]: Try creating `~/.zshenv` with the following content:

```zsh
alias -g -- -L='| less'
```

Aliases of this kind aren't unheard of. For example: https://github.com/ohmyzsh/ohmyzsh/blob/dc59c661ac1e6c7f02382a50008470f9a6eaf636/plugins/common-aliases/common-aliases.plugin.zsh#L24-L33

Disabling aliases is a standard practice when writing code that's meant to be executed within interactive zsh.

**Comment by segrey** [comment]
@romkatv I'm sorry for being slow. Thank you very much for the PR!

**Comment by segrey** [comment]
@romkatv I really appreciate your help. Merged manually as 80b99895f9f6c526011250da70530886c11cd3b0.

**Comment by romkatv** [comment]
> @romkatv I'm sorry for being slow. Thank you very much for the PR!

No worries. Thanks for taking care of it!

**Pull request closed without merging**